File: LeetCode/198.php
<?php
class Solution
{
    /**
     * @param Integer[] $nums
     * @return Integer
     */
    function massage($nums)
    {
        $count = count($nums);
        if ($count < 2)
            return empty($nums) ? 0 : $nums[0];
        $temp[0] = $nums[0];
        $temp[1] = max($nums[0], $nums[1]);
        for ($i = 2; $i < $count; $i++) {
            $temp[$i] = max($temp[$i - 1], $temp[$i - 2] + $nums[$i]);
        }
        return $temp[$count - 1];
    }

    function rob($nums)
    {
        $count = count($nums);
        $a = $b = 0;
        for ($i = 0; $i < $count; $i++) {
            $temp = max($b, $a + $nums[$i]);
            $a = $b;
            $b = $temp;
        }
        return $b;
    }
}

File: LeetCode/198.java
class Solution
{
    public int rob(int[] nums) {
        int a = 0, b = 0, temp = 0;
        for (int i = 0, len = nums.length; i < len; i++) {
            temp = Math.max(b, a + nums[i]);
            a = b;
            b = temp;
        }
        return b;
    }

    public int rob1(int[] nums) {
        int n = nums.length;
        if (n == 0) { return 0; }
        if (n == 1) { return nums[0]; }
        int[] dp = new int[n];
        dp[0] = nums[0];
        dp[1] = Math.max(nums[0], nums[1]);
        for (int i = 2; i < n; i++) {
            dp[i] = Math.max(dp[i - 1], dp[i - 2] + nums[i]);
        }
        return dp[n - 1];
    }
}
